MWI: Add lib/boundkeypair for bound keypair backend implementation

timothyb89 opened the pull request:
This splits the backend implementation of lib/boundkeypair out from the main bound keypair joining PR in #54371. It makes no changes beyond those already reviewed in that PR.

timothyb89 committed:
MWI: Add lib/boundkeypair for bound keypair backend implementation

This splits the backend implementation of lib/boundkeypair out from
the main bound keypair joining PR in #54371. It makes no changes
beyond those already reviewed in that PR.

timothyb89 commented:
Reviewer note, this just exists to reduce the changed LoC count in https://github.com/gravitational/teleport/pull/54371 and to appease the size check. This package is relatively isolated and could be split out of the larger PR without much effort.

zmb3 commented on lib/boundkeypair/bound_keypair.go:
```diff
@@ -0,0 +1,176 @@
+	return nil
+}
+
+// RecoveryModes returns a list of all supported recovery modes
+func RecoveryModes() []RecoveryMode {
+	return []RecoveryMode{
+		RecoveryModeStandard,
+		RecoveryModeRelaxed,
+		RecoveryModeInsecure,
+	}
+}
```
Consider moving this block up next to where the various modes are declared so it's easier to keep them in sync.

zmb3 commented on lib/boundkeypair/bound_keypair_test.go:
```diff
@@ -0,0 +1,266 @@
+)
+
+func newTestKeypair(t *testing.T) crypto.Signer {
+	key, err := cryptosuites.GenerateKey(context.Background(), func(ctx context.Context) (types.SignatureAlgorithmSuite, error) {
```
Many of these helper funcs could do with a `t.Helper()` call.

```suggestion
	t.Helper()
	key, err := cryptosuites.GenerateKey(context.Background(), func(ctx context.Context) (types.SignatureAlgorithmSuite, error) {
```

timothyb89 replied:
Ah, good call. Fixed now, thanks!

zmb3 commented on lib/boundkeypair/experiment/experiment.go:
```diff
@@ -0,0 +1,40 @@
+
+var mu sync.Mutex
+
+var experimentEnabled = os.Getenv("TELEPORT_BOUND_KEYPAIR_JOINING_EXPERIMENT") == "1"
```
Can we use `strconv.ParseBool` here just to be consistent with other similar env vars?

You might also consider a `TELEPORT_UNSTABLE_` prefix for the var name to be consistent with similar approaches we've taken in the past. This communicates to the person who enables the feature that it's early and subject to change in the future.

timothyb89 replied:
I've changed the impl to use `strconv.ParseBool()`. We've been copying and pasting this for our feature flags for a while and I see why we went with the simpler implementation for these hacky flags, handling the error case does make things a bit uglier.

zmb3 replied:
I typically don't handle the error case - if you get an error from `ParseBool` you are guaranteed to get `false` (not enabled).

timothyb89 replied:
Ah, good call, that's much simpler. Thanks!

zmb3 commented on lib/boundkeypair/experiment/experiment.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// SetEnabled sets the experiment enabled flag.
+func SetEnabled(enabled bool) {
```
Do we need to be able to change this at runtime? It might be simpler to check the env var on startup and require a process restart to pick up changes.

zmb3 commented on lib/boundkeypair/experiment/experiment.go:
```diff
@@ -0,0 +1,40 @@
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package experiment
```
I don't love the name of the package. At the call site, it's going to look something like `if experiment.Enabled() { ...` and the first question the reader will have is "what experiment?"

timothyb89 replied:
That's fair enough, I've renamed the package.

timothyb89 committed:
Apply code review suggestions

- Renames the experiment package
- Fixes missing test helper flags
- Renames the experiment env var
- Uses strconv.ParseBool()

timothyb89 committed:
Simplify experiment flag
